### PR TITLE
add mean

### PR DESCRIPTION
**Pull request by hj-wei**

### CODE REVIEW

File: lib/transforms/onnxextension_legalizer.cc [hunk 1]
```diff
@@ -392,6 +392,28 @@ static std::vector<Def> ConvertMaximum(const ONNXExtensionInst* ext,
...
+static std::vector<Def> ConvertMean(const ONNXExtensionInst* ext,
+                                    IRBuilder* builder) {
+  // Conver to a chain of Mean.
+  auto input = ext->GetOperand(0);
+  auto n = ext->GetNumOfOperands();
+  ConstantBuilder c_builder(ext->GetParent()->GetParent());
+  HLCHECK(n >= 1);
+  if (n == 1) {
+    return {ext->GetOperand(0)};
+  }
```
Comment (weimingzha0) [inline lib/transforms/onnxextension_legalizer.cc:404]: if (n==2) , use original mean?
for n >= 3, can we do 
```
t = concat(unsqueeze(a, 0), unsqueeze(b, 0), unsqueeze(c, 0 ...)
reduce_mean(t, axis = 0, keep_dim = false)
```
?

**Review by weimingzha0 — APPROVED** [review]

File: lib/transforms/onnxextension_legalizer.cc (hunk 1 above)
Comment (weimingzha0) [reply lib/transforms/onnxextension_legalizer.cc:404]: chained "add" is ok

File: lib/transforms/onnxextension_legalizer.cc [hunk 2]
```diff
@@ -392,6 +392,28 @@ static std::vector<Def> ConvertMaximum(const ONNXExtensionInst* ext,
...
+  auto op0 = builder->CreateAdd(ext->GetName(), input, ext->GetOperand(1));
+  for (unsigned i = 2; i < n; ++i) {
+    op0 = builder->CreateAdd(ext->GetName() + std::to_string(i - 1), *op0,
+                             ext->GetOperand(i));
+  }
+  Def mean_div =
+      *c_builder.CreateConstant("NumOfOperands", halo::Type{DataType::FLOAT32},
```
Comment (weimingzha0) [inline lib/transforms/onnxextension_legalizer.cc:411]: mean_div should has the same type as input ("All inputs and outputs must have the same data type.")

input type could be float16, bf16, double, float

for "div" , I think it requires all input be the same type.

Comment (hj-wei) [reply lib/transforms/onnxextension_legalizer.cc:411]: Ok,I'll fix it

File: lib/transforms/onnxextension_legalizer.cc [hunk 3]
```diff
@@ -392,6 +392,56 @@ static std::vector<Def> ConvertMaximum(const ONNXExtensionInst* ext,
...
+  // Conver to a chain of Mean.
+  auto n = ext->GetNumOfOperands();
+  ConstantBuilder c_builder(ext->GetParent()->GetParent());
+  HLCHECK(n >= 1);
+  if (n == 1) {
+    return {ext->GetOperand(0)};
+  }
+  auto op0 = builder->CreateAdd(ext->GetName(), ext->GetOperand(0),
+                                ext->GetOperand(1));
+  for (unsigned i = 2; i < n; ++i) {
```
Comment (weimingzha0) [inline lib/transforms/onnxextension_legalizer.cc:406]: for n == 2, can we use original "mean"?
Regular mean (2 operands) are commonly used, we don't want to convert to "add & div". that's not efficient. For more than 2 operands, we can work around it.

File: include/halo/lib/ir/onnx_convert.td [hunk 4]
```diff
@@ -311,6 +311,8 @@ def ONNX_MaxPool : OpMapping<"MaxPool", PoolingMax> {
     ExtensionAttr<"storage_order", Integer, "0">];
 }
 
+def ONNX_Mean : ONNXExtension<"Mean">;
```
Comment (weimingzha0) [inline include/halo/lib/ir/onnx_convert.td:314]: Use OpMapping<"Mean", Mean> ?

**Pull request merged**